File: Divide and Conquer Algorithm/Merge Sort/mergeSort.c
// Merge Sort (DAC/Recursive Approach)
#include <stdio.h>
#include <malloc.h>



// function for Merging Arrays
// A – input array, low – lower index, high – higher index
void Merge(int arr[], int low, int mid, int high)
{

    int i=low, j=mid+1, k=low, B[high - low + 1]; // B is a temporary array

    // k used for indexing array b;
    // i and j used for indexing on array A

    while (i <= mid && j <= high) {
        if(arr[i] <= arr[j]) {
            B[k]=arr[i];
            k=k+1;
            i=i+1;
        }
        else {
            B[k]=arr[j];
            k=k+1;
            j=j+1;
        }
    }

    while (i <= mid) { // Copy the remaining elements in first half of A to B
        B[k] = arr[i];
        k = k + 1;
        i = i + 1;
    }


    while (j <= high) { // Copy the remaining elements in second half of A to B
        B[k] = arr[j];
        k = k + 1;
        j = j + 1;
    }

    //copy the sorted elements into array A into its respective positions from Array B
    for(k=low; k <= high; k++) {
        arr[k] = B[k];
    }
    free(B);
}



// function for MergeSort
void MergeSort(int A[], int low, int high) // A – input array, low – lower index, high – higher index
{
    if (low < high) // the array has at least two elements
    {
        int mid = (low + high) / 2;

        // call the merge sort on first half of the elements of A
        MergeSort(A, low, mid);


        //  call the merge sort on second half of the elements of A
        MergeSort(A, mid+1, high);

        // Merge the sorted elements of the first half and second half
        Merge(A, low, mid, high);

    }
}


/* Function to print an array */
void printArray(int A[], int size)
{
    int i;
    for (i = 0; i < size; i++)
        printf(" %d ", A[i]);
    printf("\n");
}



int main()
{
    int size;
    printf("\n Enter the number of elements in the array: ");
    scanf("%d", &size);

    int arr[size];

    printf("\n Enter the elements of the array :\n");
    for(int i=0; i<size; i++) {
        scanf("%d", &arr[i]);
    }

    printf("\n Entered Array : " );
    printArray(arr, size);

    MergeSort(arr, 0, size-1);

    printf("\n Sorted array : ");
    printArray(arr, size);

    return 0;
}

File: Divide and Conquer Algorithm/Merge Sort/mergeSort.py
# Python program for implementation of MergeSort

# Merges two subarrays of arr[].
# First subarray is arr[l..m]
# Second subarray is arr[m+1..r]
def merge(arr, low, mid, high):
    n1 = mid - low + 1
    n2 = high - mid

    # create temp arrays
    L = [0] * (n1)
    R = [0] * (n2)

    # Copy data to temp arrays L[] and R[]
    for i in range(0, n1):
        L[i] = arr[low + i]

    for j in range(0, n2):
        R[j] = arr[mid + 1 + j]

        # Merge the temp arrays back into arr[low...high]
    i = 0  # Initial index of first subarray
    j = 0  # Initial index of second subarray
    k = low # Initial index of merged subarray

    while (i < n1 and j < n2):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1

    # Copy the remaining elements of L[], if there
    # are any
    while (i < n1):
        arr[k] = L[i]
        i += 1
        k += 1

    # Copy the remaining elements of R[], if there
    # are any
    while (j < n2):
        arr[k] = R[j]
        j += 1
        k += 1


# low is for left index and high is right index of the sub-array of arr to be sorted
def mergeSort(arr, low, high):
    if (low < high):
        mid = (low + high) // 2

        # Sort first and second halves
        mergeSort(arr, low, mid)
        mergeSort(arr, mid + 1, high)
        merge(arr, low, mid, high)


arr_size = int(input('Enter the input array size : '))
arr = []

print('Enter the elements of array :')
for i in range(arr_size):
    input_data = int(input())
    arr.append(input_data)

print('Input array :',arr)

mergeSort(arr, 0, arr_size - 1)

print('\n\nSorted array :', arr)
